samples/nrf9160/modem_shell/src/location/location_shell.c

```diff
@@ -0,0 +1,401 @@
+	}
+
+	return service;
+}
+
+/******************************************************************************/
+
+void location_ctrl_event_handler(const struct location_event_data *event_data)
+{
+	char gmaps_str[512];
+
+	switch (event_data->id) {
+	case LOCATION_EVT_LOCATION:
+		mosh_print("Location:");
+		mosh_print(
+			"  used method: %s (%d)",
+			location_method_str(event_data->location.method),
+			event_data->location.method);
+		mosh_print("  latitude: %.06f", event_data->location.latitude);
+		mosh_print("  longitude: %.06f", event_data->location.longitude);
+		mosh_print("  accuracy: %.01f m", event_data->location.accuracy);
+		if (event_data->location.datetime.valid) {
+			mosh_print(
+				"  date: %04d-%02d-%02d",
+				event_data->location.datetime.year,
+				event_data->location.datetime.month,
+				event_data->location.datetime.day);
+			mosh_print(
+				"  time: %02d:%02d:%02d.%03d UTC",
+				event_data->location.datetime.hour,
+				event_data->location.datetime.minute,
+				event_data->location.datetime.second,
+				event_data->location.datetime.ms);
+		}
+		sprintf(gmaps_str, "  Google maps URL: https://maps.google.com/?q=%f,%f",
+			event_data->location.latitude, event_data->location.longitude);
+		mosh_print("%s", gmaps_str);
+		break;
+
+	case LOCATION_EVT_TIMEOUT:
+		mosh_error("Location request timed out");
+		break;
+
+	case LOCATION_EVT_ERROR:
+		mosh_error("Location request failed");
+		break;
+	case LOCATION_EVT_GNSS_ASSISTANCE_REQUEST:
+#if defined(CONFIG_LOCATION_METHOD_GNSS_AGPS_EXTERNAL)
+		mosh_print("MoSh: A-GPS request from Location library (ephe: 0x%08x alm: 0x%08x flags: 0x%02x)",
+		event_data->request.sv_mask_ephe,
+		event_data->request.sv_mask_alm,
+		event_data->request.data_flags);
+#endif
+		break;
+	}
+}
+
+void location_ctrl_init(void)
+{
+	int ret;
+
+	ret = location_init(location_ctrl_event_handler);
+	if (ret) {
+		mosh_error("Initializing the Location library failed, err: %d\n", ret);
+	}
+}
+
+/******************************************************************************/
+
+int location_shell(const struct shell *shell, size_t argc, char **argv)
+{
+	struct location_shell_cmd_args loc_cmd_args = {
+		.command = LOCATION_CMD_NONE
+	};
+
+	int interval = 0;
+	bool interval_set = false;
+
+	int gnss_timeout = 0;
+	bool gnss_timeout_set = false;
+
+	enum location_accuracy gnss_accuracy = 0;
+	bool gnss_accuracy_set = false;
+
+	int cellular_timeout = 0;
+	bool cellular_timeout_set = false;
+	enum location_service cellular_service = LOCATION_SERVICE_ANY;
+
+	int wifi_timeout = 0;
+	bool wifi_timeout_set = false;
+	enum location_service wifi_service = LOCATION_SERVICE_ANY;
+
+	enum location_method method_list[CONFIG_LOCATION_METHODS_LIST_SIZE] = { 0 };
+	int method_count = 0;
+
+	int opt;
+	int ret = 0;
+	int long_index = 0;
+
+	if (argc < 2) {
+		goto show_usage;
+	}
+
+	/* command = argv[0] = "location" */
+	/* sub-command = argv[1] */
+	if (strcmp(argv[1], "get") == 0) {
```

Comment:
As discussed yesterday, MOSH has a mixture of ways to do the parsing with COMMAND/SUBCOMMAND macros, strcmp and getopt(). This would need a wider re-design which has been on our mind but being honest, I don't know if we ever get there. Auto-complete with tab would be a good motivation for it though.